src/freenet/clients/http/FirstTimeWizardNewToadlet.java

```diff
@@ -0,0 +1,329 @@
+            password = request.getPartAsStringFailsafe("password", SecurityLevelsToadlet.MAX_PASSWORD_LENGTH + 1);
+            String passwordConfirmation = request.getPartAsStringFailsafe("confirmPassword", SecurityLevelsToadlet.MAX_PASSWORD_LENGTH);
+
+            // validate
+            if (haveMonthlyLimit.isEmpty()) {
+                try {
+                    long downloadLimit = this.downloadLimit.isEmpty() ? 0 : Fields.parseInt(this.downloadLimit + "KiB");
+                    if (downloadLimit < Node.getMinimumBandwidth()) {
+                        errors.put("downloadLimitError",
+                                l10n("valid.downloadLimit", Integer.toString(Node.getMinimumBandwidth() / KiB)));
+                    }
+                } catch (NumberFormatException e) {
+                    errors.put("downloadLimitError",
+                            l10n("valid.number.prefix.downloadLimit") + " " + e.getMessage());
+                }
+
+                try {
+                    long uploadLimit = this.uploadLimit.isEmpty() ? 0 : Fields.parseInt(this.uploadLimit + "KiB");
+                    if (uploadLimit < Node.getMinimumBandwidth()) {
+                        errors.put("uploadLimitError",
+                                l10n("valid.uploadLimit", Integer.toString(Node.getMinimumBandwidth() / KiB)));
+                    }
+                    int nanosInSecond = (int) SECONDS.toNanos(1);
+                    if (nanosInSecond < uploadLimit) { // see Node set outputBandwidthLimit
+                        errors.put("uploadLimitError",
+                                l10n("valid.uploadLimitMax", Integer.toString(nanosInSecond / KiB)));
+                    }
+                } catch (NumberFormatException e) {
+                    errors.put("uploadLimitError",
+                            l10n("valid.number.prefix.uploadLimit") + " " + e.getMessage());
+                }
+            } else {
+                try {
+                    double monthlyLimit = 0;
+                    if (!bandwidthMonthlyLimit.isEmpty()) {
+                        monthlyLimit = Double.parseDouble(bandwidthMonthlyLimit);
+                    }
+                    if (monthlyLimit < BandwidthLimit.minMonthlyLimit) {
+                        errors.put("bandwidthMonthlyLimitError",
+                                l10n("valid.bandwidthMonthlyLimit", String.format("%.2f", BandwidthLimit.minMonthlyLimit)));
+                    }
+                } catch (NumberFormatException e) {
+                    errors.put("bandwidthMonthlyLimitError",
+                            l10n("valid.number.prefix.bandwidthMonthlyLimit") + " " + e.getMessage());
+                }
+            }
+
+            try {
+                long storageLimit = this.storageLimit.isEmpty() ? 0 : Fields.parseLong(this.storageLimit + "GiB");
+                if (storageLimit < Node.MIN_STORE_SIZE * 5 / 4) { // min store size + 10% for client cache + 10% for slashdot cache
```

Comment:
This 5/4 business is fair enough because it's explained in the comment, but it seems strange to bury logic like this in form validation code. What do you think about putting it in `DatastoreUtil` as `minStorageLimit` where it'll be alongside `autodetectDatastoreSize()` and `maxDatastoreSize()`?

Reply:
damn, I removed the comment when putting it into a constant. I recovered the comment in a hopefully final commit :-)

I would keep it in a constant of the toadlet for the time being. If we need it at another place, we can still pull it into the storage util.